### PR TITLE
Add documentation for lambda functions

### PR DESCRIPTION
**Pull request by haozhun**

### CODE REVIEW

**Comment by haozhun** [comment]
@electrum I have two questions for you:

* I find it useful to include types when documenting lambda functions. Is that ok?
* Having `)` in the `.. function` line messes formatting up. Is there a workaround?

File: presto-docs/src/main/sphinx/functions/lambda.rst [hunk 1]
```diff
@@ -0,0 +1,67 @@
+============================
+Lambdas and Lambda Functions
+============================
+
+Lambda Expression
+-----------------
+
+Lambda expressions in Presto are written with ``->``. For example::
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/lambda.rst]: On all of these, you can remove the "For example:" and just do the colon at the end of the previous sentence. See JSON functions for example.

File: presto-docs/src/main/sphinx/functions/lambda.rst [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+* Subqueries are not supported.
+
+Lambda Functions
+----------------
+
+.. function:: transform(array ARRAY(T), function FUNCTION(T,U)) -> ARRAY(U)
+
+    Returns an array that applies ``function`` to each element of ``array``. For example::
+
+        transform(ARRAY [], x -> x + 1) -> ARRAY []
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/lambda.rst]: Make all these `SELECT` queries with the output at end with a comment. Using an arrow is really confusing given that it's also part of the lambda syntax which is being demonstrated. See JSON functions for examples.

File: presto-docs/src/main/sphinx/functions/lambda.rst [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+.. function:: filter(array ARRAY(T), function FUNCTION(T,BOOLEAN)) -> ARRAY(T)
+
+    Constructs an array from those elements of ``array`` for which ``function`` returns true. For example::
+
+        filter(ARRAY [], x -> true) -> ARRAY []
+        filter(ARRAY [5, -6, NULL, 7], x -> x > 0) -> ARRAY [5, 7]
+        filter(ARRAY [5, NULL, 7, NULL], x -> x IS NOT NULL) -> ARRAY [5, 7]
+        transform(ARRAY [ARRAY [1, NULL, 2], ARRAY[3, NULL]], a -> filter(a, x -> x IS NOT NULL)) -> ARRAY [ARRAY [1, 2], ARRAY [3]]
+
+.. function:: reduce(array ARRAY(T), initialState S, inputFunction FUNCTION(T,S,S), outputFunction FUNCTION(S,R)) -> R
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/lambda.rst]: I agree having types is useful, but I find this hard to read. Let's see if we can improve this. Maybe breaking out the types into the description?

screen shot 2016-12-07 at 10 04 43 am

File: presto-docs/src/main/sphinx/functions/lambda.rst [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    ``outputFunction`` will be invoked to turn the final state into the result value. It may be identity function (``i -> i``).
+    For example::
+
+        reduce(ARRAY [], 0, (x, s) -> x + s, s -> s) -> 0
+        reduce(ARRAY [5, 20, 50], 0, (x, s) -> x + s, s -> s) -> 75
+        reduce(ARRAY [5, 20, NULL, 50], 0, (x, s) -> x + s, s -> s) -> NULL
+        reduce(ARRAY [5, 20, NULL, 50], 0, (x, s) -> COALESCE(x, 0) + s, s -> s) -> 75
+        reduce(ARRAY [5, 20, NULL, 50], 0, (x, s) -> IF(x IS NULL, s, x + s), s -> s) -> 75
+        reduce(ARRAY [2147483647, 1], CAST (0 AS BIGINT), (x, s) -> x + s, s -> s) -> 2147483648
+        reduce( -- this reduction calculates arithmetic average
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/lambda.rst]: Let's write it like this:
```sql
SELECT reduce(ARRAY [5, 6, 10, 20],
              CAST(ROW(0.0, 0) AS ROW(sum DOUBLE, count INTEGER)),
              (x, s) -> CAST(ROW(x + s.sum, s.count + 1) AS ROW(sum DOUBLE, count INTEGER)),
              s -> IF(s.count = 0, NULL, s.sum / s.count)));
-- calculate arithmetic average: 10.25

File: presto-docs/src/main/sphinx/functions/lambda.rst [hunk 5]
```diff
@@ -0,0 +1,76 @@
+============================
+Lambdas and Lambda Functions
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/lambda.rst]: Lambda Expressions and Functions

File: presto-docs/src/main/sphinx/functions/lambda.rst [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    (x, y) -> x + y
+    x -> regexp_like(x, 'a+')
+    x -> x[1] / x[2]
+    x -> IF(x > 0, x, -x)
+    x -> COALESCE(x, 0)
+    x -> CAST(x AS JSON)
+
+Most SQL expressions can be used in a lambda body, with a few exceptions:
+
+* The ``TRY`` function is not supported yet. (:func:`TRY_CAST` is supported.)
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/lambda.rst]:
```
* :func:`try` is not supported yet. (:func:`try_cast` is supported.)

File: presto-docs/src/main/sphinx/functions/lambda.rst [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+* Capture is not supported yet:
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/lambda.rst]: Add a newline before this item, too, so they are consistent.

File: presto-docs/src/main/sphinx/functions/lambda.rst (hunk 6 above)
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/lambda.rst]:
```
* Subqueries are not supported.
* :func:`try` is not supported yet. (:func:`try_cast` is supported.)
* Capture is not supported yet:

  * Columns or relations cannot be referenced.
  * Only lambda variables from the inner-most lambda expression can be referenced.

File: presto-docs/src/main/sphinx/functions/lambda.rst [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        SELECT map_filter(MAP(ARRAY['k1', 'k2', 'k3'], ARRAY[20, 3, 15]), (k, v) -> v > 10); -- {k1 -> 20, k3 -> 15}
+
+.. function:: transform(array ARRAY<T>, function FUNCTION<T,U>) -> ARRAY<U>
+
+    Returns an array that applies ``function`` to each element of ``array``::
+
+        SELECT transform(ARRAY [], x -> x + 1); -- []
+        SELECT transform(ARRAY [5, 6], x -> x + 1); -- [6, 7]
+        SELECT transform(ARRAY [5, NULL, 6], x -> COALESCE(x, 0) + 1); -- [6, 1, 7]
+        SELECT transform(ARRAY ['x', 'abc', 'z'], x -> CONCAT(x, '0')); -- ['x0', 'abc0', 'z0']
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/lambda.rst]: Use the standard concat operator `||`

File: presto-docs/src/main/sphinx/functions/lambda.rst [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        SELECT reduce(ARRAY [], 0, (x, s) -> x + s, s -> s); -- 0
+        SELECT reduce(ARRAY [5, 20, 50], 0, (x, s) -> x + s, s -> s); -- 75
+        SELECT reduce(ARRAY [5, 20, NULL, 50], 0, (x, s) -> x + s, s -> s); -- NULL
+        SELECT reduce(ARRAY [5, 20, NULL, 50], 0, (x, s) -> COALESCE(x, 0) + s, s -> s); -- 75
+        SELECT reduce(ARRAY [5, 20, NULL, 50], 0, (x, s) -> IF(x IS NULL, s, x + s), s -> s); -- 75
+        SELECT reduce(ARRAY [2147483647, 1], CAST (0 AS BIGINT), (x, s) -> x + s, s -> s); -- 2147483648
+        SELECT reduce(ARRAY [5, 6, 10, 20],
+                      CAST(ROW(0.0, 0) AS ROW(sum DOUBLE, count INTEGER)),
+                      (x, s) -> CAST(ROW(x + s.sum, s.count + 1) AS ROW(sum DOUBLE, count INTEGER)),
+                      s -> IF(s.count = 0, NULL, s.sum / s.count))
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/lambda.rst]: Add semicolon at end

Comment (electrum) [inline presto-docs/src/main/sphinx/functions/lambda.rst]: Align the output:
```
              s -> IF(s.count = 0, NULL, s.sum / s.count));
              -- calculates arithmetic average: 10.25

File: presto-docs/src/main/sphinx/functions/lambda.rst [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+        SELECT transform(ARRAY [], x -> x + 1); -- []
+        SELECT transform(ARRAY [5, 6], x -> x + 1); -- [6, 7]
+        SELECT transform(ARRAY [5, NULL, 6], x -> COALESCE(x, 0) + 1); -- [6, 1, 7]
+        SELECT transform(ARRAY ['x', 'abc', 'z'], x -> CONCAT(x, '0')); -- ['x0', 'abc0', 'z0']
+        SELECT transform(ARRAY [ARRAY [1, NULL, 2], ARRAY[3, NULL]], a -> filter(a, x -> x IS NOT NULL)); -- [[1, 2], [3]]
+
+.. function:: reduce(array ARRAY<T>, initialState S, inputFunction FUNCTION<T,S,S>, outputFunction FUNCTION<S,R>) -> R
+
+    Returns a single value reduced from the ``array``. ``inputFunction`` will
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/lambda.rst]: Remove "the"

**Pull request merged**